### PR TITLE
Fixed deletion of images with id greater than 1000

### PR DESCRIPTION
**Pull request by MaksFM**
Inside the $this->getReferenceImage() method, the id of the $media object is used, but at the time of the postRemove event, the $media object is detached from an EntityManager and its id is null. During the postRemove event, the $media->getId() method returns null, and the $this->getReferenceImage() method returns the wrong path and the file remains in the file system as garbage. My changes fix this.

## Subject

I am targeting this branch, because I use the last stable release in my project, and found an error in it, but my changes are backward compatible and fix it..

## Changelog

```markdown
### Fixed
Original image files with id greater than 1000 are not deleted from the file system.
```

### CODE REVIEW

**Comment by MaksFM** [comment]
Demonstrates incorrect work of the getReferenceImage() method during the postRemove event. Details in the comments of the attached file. 

[tests.zip](https://github.com/sonata-project/SonataMediaBundle/files/3165988/tests.zip)

This is my first contribution, I don’t know how to provide the necessary tests more correctly.

**Comment by MaksFM** [comment]
Good afternoon, I added tests for the code that fixes the described error.  Please give your comments if this solution is not suitable for the general repository.  Just then I will use in my project a private solution of the problem, not affecting the internal sources of the corresponding component.  Although, I still consider the most preferred bug fix in the main repository.

File: src/Provider/BaseProvider.php [hunk 1]
```diff
@@ -63,6 +63,11 @@ abstract class BaseProvider implements MediaProviderInterface
      */
     protected $name;
 
+    /**
+     * @var array
+     */
+    protected $clones = [];
```
Comment (core23) [inline src/Provider/BaseProvider.php]: Please make this private by default

Comment (MaksFM) [reply src/Provider/BaseProvider.php]: Done.

File: src/Provider/BaseProvider.php [hunk 2]
```diff
@@ -63,6 +63,11 @@ abstract class BaseProvider implements MediaProviderInterface
      */
     protected $name;
 
+    /**
+     * @var array
```
Comment (phansys) [inline src/Provider/BaseProvider.php]:
```suggestion
     * @var MediaInterface[]
```

File: src/Provider/BaseProvider.php [hunk 3]
```diff
@@ -191,6 +199,13 @@ public function preRemove(MediaInterface $media)
      */
     public function postRemove(MediaInterface $media)
     {
+        $hash = spl_object_hash($media);
+
+        if (array_key_exists($hash, $this->clones)) {
```
Comment (phansys) [inline src/Provider/BaseProvider.php]:
```suggestion
        if (isset($this->clones[$hash])) {
```

File: tests/Provider/BaseProviderTest.php [hunk 4]
```diff
@@ -151,9 +159,9 @@ public function getAbsolutePath(MediaInterface $media): void
     /**
      * {@inheritdoc}
      */
-    public function getReferenceImage(MediaInterface $media): void
+    public function getReferenceImage(MediaInterface $media)
```
Comment (phansys) [inline tests/Provider/BaseProviderTest.php]:
```suggestion
    public function getReferenceImage(MediaInterface $media): string
```

File: tests/Provider/BaseProviderThisVerOrBelowTest.php [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+        return $provider;
+    }
+
+    public function testPostRemove(): void
+    {
+        $provider = $this->getProvider();
+        $media = new Media();
+        
+        //The ID is more than a thousand, the bug manifests itself under this condition.
```
Comment (core23) [inline tests/Provider/BaseProviderThisVerOrBelowTest.php]: You can remove this comment, as the context of *the bug* is missing after merging this PR.

Comment (MaksFM) [reply tests/Provider/BaseProviderThisVerOrBelowTest.php]: This test is more likely to demonstrate the error before correction and its absence after. In the resulting branch, it is preferable to delete it completely.

File: tests/Provider/BaseProviderThisVerOrBelowTest.php [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    public function getReferenceImage(MediaInterface $media): string
+    {
+        // A copy of the code from \Sonata\MediaBundle\Provider\FileProvider::getReferenceImage()
+        $path = sprintf('%s/%s',
+            $this->generatePath($media),
+            $media->getProviderReference()
+        );
+        
+        // Test code.
+        AbstractProviderTest::assertSame('/0001/02/1f981a048e7d8b671415d17e9633abc0059df394.png', $path);
```
Comment (phansys) [inline tests/Provider/BaseProviderThisVerOrBelowTest.php]: `AbstractProviderTest` => `self`.
BTW, fixture methods should avoid the testing responsibility. The expected path should not be hardcoded, it should be provided by context (argument or so).

File: tests/Provider/BaseProviderTest.php [hunk 7]
```diff
@@ -96,10 +96,37 @@ public function testMetadata(): void
...
+        $this->assertArrayNotHasKey($hash, $prop->getValue($provider));
+        $this->assertSame('/0001/02/1f981a048e7d8b671415d17e9633abc0059df394.png', $provider->prevReferenceImage);
+
+        $prop->setAccessible(false);
+    }
 }
 
 class TestProvider extends BaseProvider
 {
+    public $prevReferenceImage;
```
Comment (phansys) [inline tests/Provider/BaseProviderTest.php]:
```suggestion
    /**
     * @var string
     */
    public $prevReferenceImage;
```

**Comment by core23** [comment]
FlintCI is failing. Can you please apply the cs fixes @MaksFM

**Comment by phansys** [comment]
Could you please squash your commits?

**Comment by core23** [comment]
Thanks @MaksFM

**Pull request merged**